### ISSUE TITLE
Combining Datasets

### ISSUE THREAD

**Issue by DinitoFate** [ultralytics/ultralytics#7845]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I have been trying to combine datasets for my project, trying to save some time. I have my own annotated class, which i put the work in, but i am trying to save time by getting the second class already done and download the dataset, which i successfully did. Now, my problem consists of merging classes. When I try to merge them in roboflow, I manage less successfully. Both of my project merge and the classes seem different in roboflow, however, when i am trying to train the merged dataset, all label files and their positions start with 0, which, please correct me if I am wrong, indicates that the (0th) first class is used, as before the merger, both datasets had only one class.
The second dataset that I am trying to use has twice as many pictures, could that also be causing this problem? I tried managing classes and also modify them in roboflow, things i found on the internet that were supposed to help. 
I also tried varying epoch size, no difference.

Here is the training command that i've been using : 
yolo task=detect mode=train epochs=3 data=data.yaml model=yolov8m.pt imgsz=640.

Thanks for all the help and advice.

**Comment by glenn-jocher** [comment]
@DinitoFate hello! It sounds like you're encountering an issue with class labels after merging datasets. When combining datasets with different classes, it's crucial to ensure that the class indices in the annotation files are correctly mapped and unique across the combined dataset. If both datasets start with class index 0, they need to be adjusted so that the second dataset's classes continue where the first one left off.

For instance, if your first dataset has classes 0-4, the second dataset should start with class 5 to avoid conflicts. This remapping should be done before training to ensure the model correctly learns all distinct classes.

The issue you're facing with the label files starting with 0 suggests that the class indices might not have been remapped correctly after merging. The larger size of the second dataset shouldn't cause this problem, but it's important to have balanced classes to prevent class imbalance issues during training.

Please review your merged dataset's annotation files to confirm that each class has a unique index and that these indices are consistent across all label files. You might need to manually adjust the class indices or use a script to automate the process.

For further guidance on dataset preparation and troubleshooting, you can refer to our documentation at https://docs.ultralytics.com. Good luck with your training! 🚀

**Comment by DinitoFate** [comment]
@glenn-jocher Thanks a lot! You've made it clear to me now. But I am struggling to find any relevant source for adjusting the class indices, it would be a lifesaver if you could point me somewhere or help me with that in any way.

**Comment by glenn-jocher** [comment]
@DinitoFate I'm glad the previous explanation was helpful! To adjust class indices, you'll need to edit the annotation files of your datasets. Each annotation file contains class indices that correspond to the objects in the images. You'll want to increment the class indices in the second dataset's annotation files by the number of classes in the first dataset.

Unfortunately, I can't provide a direct code example, but the process involves:
1. Identifying the number of classes in your first dataset.
2. Incrementing the class indices in the second dataset's annotation files by that number.

You can write a simple script to parse the annotation files and add the required offset to the class indices. After adjusting the indices, ensure that your `data.yaml` file reflects the correct number of total classes.

For more detailed instructions on dataset preparation and annotation file format, please refer to our documentation at https://docs.ultralytics.com. Good luck with your project! 🌟

**Comment by DinitoFate** [comment]
I am new to all this as I am doing it for my Bachelor's theses and I have no prior experience regarding computer vision/machine learning and therefore prone to many basic mistakes. 
My problem was in data.yaml file, in which I did not have full path to train and valid files, therefore i was always using the files from the first dataset, which had only the first class. 
To sum it up, merging in roboflow and changing to full paths in data.yaml works, maybe this will help someone in the future, who could have a similar problem like I did.

**Comment by glenn-jocher** [comment]
@DinitoFate Great to hear that you've resolved the issue by updating the `data.yaml` file with the full paths to your training and validation files. This is indeed a common oversight for those new to the field. Your experience and solution will surely be helpful to others facing similar challenges. Best of luck with your Bachelor's thesis! If you have any more questions as you continue your work, feel free to reach out. Happy learning! 🎓👍
